src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/PipelineExtensions.cs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO.Pipelines;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.Infrastructure;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Internal.Http
+{
+    public static class PipelineExtensions
+    {
+        public static ValueTask<ArraySegment<byte>> PeekAsync(this IPipelineReader pipelineReader)
+        {
+            var input = pipelineReader.ReadAsync();
+            while (input.IsCompleted)
+            {
+                var result = input.GetResult();
+                try
+                {
+                    if (!result.Buffer.IsEmpty)
+                    {
+                        var segment = result.Buffer.First;
+                        var data = segment.GetArray();
+
+                        return new ValueTask<ArraySegment<byte>>(data);
+                    }
+                    else if (result.IsCompleted || result.IsCancelled)
+                    {
+                        return default(ValueTask<ArraySegment<byte>>);
+                    }
+                }
+                finally
+                {
+                    pipelineReader.Advance(result.Buffer.Start, result.Buffer.Start);
+                }
+                input = pipelineReader.ReadAsync();
+            }
+
+            return new ValueTask<ArraySegment<byte>>(pipelineReader.PeekAsyncAwaited(input));
+        }
+
+        private static async Task<ArraySegment<byte>> PeekAsyncAwaited(this IPipelineReader pipelineReader, ReadableBufferAwaitable readingTask)
+        {
+            while (true)
+            {
+                var result = await readingTask;
+
+                await AwaitableThreadPool.Yield();
+
+                try
+                {
+                    if (!result.Buffer.IsEmpty)
+                    {
+                        var segment = result.Buffer.First;
+                        return segment.GetArray();
+                    }
+                    else if (result.IsCompleted || result.IsCancelled)
+                    {
+                        return default(ArraySegment<byte>);
+                    }
+                }
+                finally
+                {
+                    pipelineReader.Advance(result.Buffer.Start, result.Buffer.Start);
+                }
+
+                readingTask = pipelineReader.ReadAsync();
+            }
+        }
+
+        public static async Task<ReadResult> ReadAsyncDispatched(this IPipelineReader pipelineReader)
+        {
+            var result = await pipelineReader.ReadAsync();
+
+            await AwaitableThreadPool.Yield();
```

Comment:
I just realized this is hella inefficient. We need to move this back into the pipe itself since it's the only one that understand if dispatch is needed or not (in the case of an already completed ReadAsync).